pkg/tekton/metapipeline/clientfactory.go

```diff
@@ -0,0 +1,295 @@
+	return os.RemoveAll(c.versionDir)
+}
+
+//DefaultImage returns the default image used for pipeline tasks if no other image is specified.
+func (c *ClientFactory) DefaultImage() string {
+	return c.defaultImage
+}
+
+// SetDefaultImage sets a default image.
+func (c *ClientFactory) SetDefaultImage(image string) {
+	c.defaultImage = image
+}
+
+// ServiceAccount returns the service account under which to execute the pipeline.
+func (c *ClientFactory) ServiceAccount() string {
+	return c.serviceAccount
+}
+
+// SetServiceAccount sets a non default service account for running the Tekton pipeline.
+func (c *ClientFactory) SetServiceAccount(serviceAccount string) {
+	c.serviceAccount = serviceAccount
+}
+
+func (c *ClientFactory) getPodTemplates(containerName string) (map[string]*corev1.Pod, error) {
+	podTemplates, err := kube.LoadPodTemplates(c.kubeClient, c.ns)
+	if err != nil {
+		return nil, err
+	}
+
+	return podTemplates, nil
+}
+
+func (c *ClientFactory) determineBranchIdentifier(pipelineType PipelineKind, pullRef PullRef) (string, error) {
+	var branch string
+	switch pipelineType {
+	case ReleasePipeline:
+		{
+			// no pull requests to merge, taking base branch name as identifier
+			branch = pullRef.baseBranch
+		}
+	case PullRequestPipeline:
+		{
+			if len(pullRef.pullRequests) == 0 {
+				return "", errors.New("pullrequest pipeline requested, but no pull requests specified")
+			}
+			branch = fmt.Sprintf("PR-%s", pullRef.PullRequests()[0].ID)
+		}
+	default:
+		{
+			branch = "unknown"
+		}
+	}
+	return branch, nil
+}
+
+func (c *ClientFactory) versionStreamURLAndRef() (string, string, error) {
+	devEnv, err := kube.GetDevEnvironment(c.jxClient, c.ns)
+	if err != nil {
+		return "", "", errors.Wrap(err, "unable to retrieve team environment")
+	}
+
+	if devEnv == nil {
+		return config.DefaultVersionsURL, config.DefaultVersionsRef, nil
+	}
+
+	teamSettings := devEnv.Spec.TeamSettings
+	url := teamSettings.VersionStreamURL
+	ref := teamSettings.VersionStreamRef
+	if url == "" {
+		url = config.DefaultVersionsURL
+	}
+	if ref == "" {
+		ref = config.DefaultVersionsRef
+	}
+
+	return url, ref, nil
+}
+
+func (c *ClientFactory) cloneVersionStreamIfNeeded() error {
+	url, ref, err := c.versionStreamURLAndRef()
+	if err != nil {
+		return err
+	}
+
+	if c.versionStreamURL != url || c.versionStreamRef != ref {
+		_ = os.RemoveAll(c.versionDir)
+		c.versionDir, err = c.cloneVersionStream(url, ref)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *ClientFactory) cloneVersionStream(url string, ref string) (string, error) {
+	dir, err := ioutil.TempDir("", "jx-version-repo-")
+	if err != nil {
+		return "", errors.Wrap(err, "unable to create temp dir for version stream")
+	}
+
+	logger.Infof("cloning version stream url: %s ref: %s into %s", url, ref, dir)
+
+	// Not using GitCLi Clone/ShallowClone atm, since it does not work with tags (HF)
+	args := []string{"clone", "--depth", "1", "--branch", ref, url, "."}
```

Comment:
I think this is introducing an indirect dependency to `git` CLI into the consumers of this client. :( This is not nice if I want to have a pure go process.

Reply:
true, but that's also the case if I use gits.ShallowClone. It also shells out a git command execution (or in fact multiple). 

What's the alternative? We don't have a Go impl of git, right? I agree in a perfect world we would just have a single go process.

Reply:
This is more my contemplation because I was thinking that we should get rid of this dependency in `Lighthouse` to reduce the docker image size.

There is `go-git` but yeah we shouldn't start to introduce a new dependency just for this.

Reply:
I like the idea to do rely less on external tools in general. However, trying to get away from calling git is definitely more than we can chew off here. 

That said, this is just another point pro `pipelinerunner` imo. If we were to keep `pipelinerunner` which would clone the version stream and I find it more reasonable that our direct controller images have git backed in. If lighthouse where to make an HTTP request to the `pipelinerunner` you would not have to care about this.

Reply:
Makes me also wonder a bit, why version stream is something which we clone a repo for. I would the version stream more envision as something living in the cluster and which I can query for versions of things. Whether this in cluster thing itself gets its state/updates from a repo is a different question, but I find the needs to clone odd.